pkg/controllers/subnetset/subnetset_controller.go

```diff
@@ -0,0 +1,261 @@
+	matchedCondition := getExistingConditionOfType(newCondition.Type, subnetset.Status.Conditions)
+
+	if reflect.DeepEqual(matchedCondition, newCondition) {
+		log.V(2).Info("conditions already match", "New Condition", newCondition, "Existing Condition", matchedCondition)
+		return false
+	}
+
+	if matchedCondition != nil {
+		matchedCondition.Reason = newCondition.Reason
+		matchedCondition.Message = newCondition.Message
+		matchedCondition.Status = newCondition.Status
+	} else {
+		subnetset.Status.Conditions = append(subnetset.Status.Conditions, *newCondition)
+	}
+	return true
+}
+
+func getExistingConditionOfType(conditionType v1alpha1.ConditionType, existingConditions []v1alpha1.Condition) *v1alpha1.Condition {
+	for i := range existingConditions {
+		if existingConditions[i].Type == conditionType {
+			return &existingConditions[i]
+		}
+	}
+	return nil
+}
+
+func (r *SubnetSetReconciler) setupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&v1alpha1.SubnetSet{}).
+		WithOptions(controller.Options{
+			MaxConcurrentReconciles: runtime.NumCPU(),
+		}).
+		Watches(&source.Kind{Type: &v1alpha1.VPC{}},
+			&VPCHandler{Client: mgr.GetClient()},
+			builder.WithPredicates(VPCPredicate)).
+		Watches(
+			&source.Kind{Type: &v1alpha1.SubnetPort{}},
+			&SubnetPortHandler{Reconciler: r},
+			builder.WithPredicates(SubnetPortPredicate)).
+		Complete(r)
+}
+
+// GarbageCollector collect Subnet which there is no port attached on it.
+// cancel is used to break the loop during UT
+func (r *SubnetSetReconciler) GarbageCollector(cancel chan bool, timeout time.Duration) {
+	ctx := context.Background()
+	log.Info("subnetset garbage collector started")
+	for {
+		select {
+		case <-cancel:
+			return
+		case <-time.After(timeout):
+		}
+		subnetSetList := &v1alpha1.SubnetSetList{}
+		err := r.Client.List(ctx, subnetSetList)
+		if err != nil {
+			log.Error(err, "failed to list SubnetSet CR")
+			continue
+		}
+		for _, subnetSet := range subnetSetList.Items {
+			if err := r.DeleteSubnetForSubnetSet(subnetSet, true); err != nil {
+				metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteFailTotal, MetricResTypeSubnetSet)
+			} else {
+				metrics.CounterInc(r.Service.NSXConfig, metrics.ControllerDeleteSuccessTotal, MetricResTypeSubnetSet)
+			}
+		}
+	}
+}
+
+func (r *SubnetSetReconciler) DeleteSubnetForSubnetSet(obj v1alpha1.SubnetSet, updataStatus bool) error {
+	nsxSubnets := r.Service.SubnetStore.GetByIndex(servicecommon.TagScopeSubnetCRUID, string(obj.GetUID()))
+	hitError := false
+	for _, subnet := range nsxSubnets {
+		portNums := len(common.ServiceMediator.GetPortsOfSubnet(*subnet.Id))
+		if portNums > 0 {
+			continue
+		}
+		if err := r.Service.DeleteSubnet(subnet); err != nil {
+			log.Error(err, "fail to delete subnet from subnetset cr", "ID", *subnet.Id)
+			hitError = true
+		}
```

Comment:
It could conflict with the code that allocates empty subnet. If https://github.com/tnqn/nsx-operator/pull/1#discussion_r1293346859 is accepted, perhaps it could implement garbage collection in `Reconcile`, it could be retriggered by always returning RequeueAfter as the GCInterval